pkg/server/ca/upstream_client.go

```diff
@@ -0,0 +1,339 @@
+	}
+}
+
+// WaitUntilMintX509CAStreamDone waits until the MintX509CA stream has stopped.
+func (u *UpstreamClient) WaitUntilMintX509CAStreamDone(ctx context.Context) error {
+	return u.mintX509CAStream.WaitUntilStopped(ctx)
+}
+
+// PublishJWTKey publishes the JWT key to the UpstreamAuthority. It maintains
+// an open stream to the UpstreamAuthority plugin to receive and append JWT key
+// updates to the bundle. The stream remains open until another call to
+// PublishJWTKey happens or the client is closed.
+func (u *UpstreamClient) PublishJWTKey(ctx context.Context, jwtKey *common.PublicKey) (_ []*common.PublicKey, err error) {
+	u.publishJWTKeyMtx.Lock()
+	defer u.publishJWTKeyMtx.Unlock()
+
+	req := &upstreamauthority.PublishJWTKeyRequest{
+		JwtKey: jwtKey,
+	}
+
+	firstResultCh := make(chan publishJWTKeyResult)
+	u.publishJWTKeyStream.Start(func(streamCtx context.Context) {
+		u.runPublishJWTKeyStream(streamCtx, req, firstResultCh)
+	})
+	defer func() {
+		if err != nil {
+			u.publishJWTKeyStream.Stop()
+		}
+	}()
+
+	select {
+	case result := <-firstResultCh:
+		return result.jwtKeys, result.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (u *UpstreamClient) runMintX509CAStream(ctx context.Context, req *upstreamauthority.MintX509CARequest, firstResultCh chan<- mintX509CAResult) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := u.c.UpstreamAuthority.MintX509CA(ctx, req)
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	x509CA, x509Roots, err := parseMintX509CAFirstResponse(resp)
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	if !u.c.UpstreamBundle {
+		// We have opted not to join the upstream PKI. The server CA should
+		// therefore be considered the root. Also, this means there is no
+		// reason to continue monitoring for bundle updates, so update the
+		// bundle with the server CA and return.
+		if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509CA[:1]); err != nil {
+			firstResultCh <- mintX509CAResult{err: err}
+			return
+		}
+		firstResultCh <- mintX509CAResult{x509CA: x509CA, done: true}
+		return
+	}
+
+	if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509Roots); err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	firstResultCh <- mintX509CAResult{x509CA: x509CA}
+
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			if err != io.EOF {
+				u.c.BundleUpdater.OnError(err, "MintX509CA stream closed prematurely")
+			}
+			return
+		}
+
+		x509Roots, err := parseMintX509CABundleUpdate(resp)
+		if err != nil {
+			u.c.BundleUpdater.OnError(err, "Invalid MintX509CA bundle update response")
+			continue
+		}
+
+		if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509Roots); err != nil {
+			u.c.BundleUpdater.OnError(err, "Failed to append upstream X.509 roots")
```

Comment:
And here `"Failed to store bundle update received from the upstream authority plugin."`

One way to read the current message here is "I failed to send my certificate to some upstream server", which is not what we're going for

Reply:
:+1: